### PR TITLE
Content link

### LINKED ISSUE #11: CourseService: Endpoint for requesting affiliated Course ID for Resource ID
## Parent
#12 

## Description
In the use case of having a request done directly to a service handling a resource, the service will have to find out if a user has access to the resource. For this, the service requests the course ID and the resource's availability (simple boolean value) of a resource from the course service.

## Acceptance Criteria
- [ ] course service provides an endpoint that accepts a resource ID and returns the course ID and its availability as a boolean.

### PR DESCRIPTION
**Pull request by BlueWaves2**
## Description of changes
Courses now have a list of resources. These resources can be contained in multiple courses. 
This PR contains an end point that allows to retrieve a list of courses with the availability of a resource for a course.
Closes #11 
  
## How has this been tested?
Please describe the test strategy you followed.

- [X] automated unit test
- [X] automated integration test

## Checklist before requesting a review
- [X] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My code fulfilles all acceptance criteria
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have made corresponding changes to the documentation
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/resources/graphql/query.graphqls [hunk 1]
```diff
@@ -20,6 +20,11 @@ type Query {
     """
     coursesById(ids: [UUID!]!): [Course!]!
 
+    """
+    Returns a set of Resource Objects for the given resource ids, containing a list of all course IDs for a resource and its availability in the course.
+    """
+    courseIdsByResource(ids: [UUID!]!): [Resource!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/query.graphqls:26]: It's a bit unintuitive that the name of the query is "courseIds..." but the return type is an array of resources. I don't really how we could call it better though

**Review by PaulBredl — CHANGES_REQUESTED** [review]
Before I look into the Java could I would first discuss the remarks I had on the schema

File: src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ResourcePk implements Serializable {
+
+    private UUID courseId;
+
+    private UUID resourceId;
+
+    public UUID getCourseId() {
```
[the remark this answers is not in the document]
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java:23]: wtf I just tagged a user named Data

File: src/main/resources/graphql/resource.graphqls [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    A list of courses a resource is present in, including its availability
+    """
+    courses: [courseIdAvailabilityMap!]!
+}
+
+type courseIdAvailabilityMap {
+    """
+    Course ID
+    """
+    course_id: UUID!
```
[the remark this answers is not in the document]
Comment (myluki2000) [reply src/main/resources/graphql/resource.graphqls:19]: We have a guideline, see the wiki.

File: src/main/resources/graphql/resource.graphqls [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    UUID of the resource.
+    """
+    resource_id: UUID!
+    """
+    A list of courses a resource is present in, including its availability
+    """
+    courses: [courseIdAvailabilityMap!]!
+}
+
+type courseIdAvailabilityMap {
```
Comment (myluki2000) [inline src/main/resources/graphql/resource.graphqls:15]: This needs to be PascalCase according to our guidelines.

Comment (myluki2000) [reply src/main/resources/graphql/resource.graphqls:15]: I would also not name this "Map" as it isn't a map. Maybe just CourseAvailabilityInformation or sth similar?

File: src/main/resources/graphql/resource.graphqls [hunk 5]
```diff
@@ -0,0 +1,24 @@
+"""
+Resources are all types of content present in a course. Each resource can be available or unavailable in a course.
+"""
+type Resource {
```
Comment (myluki2000) [inline src/main/resources/graphql/resource.graphqls:4]: I'm not completely happy with just naming this "Resource", as this type cannot be used to actually fetch or do anything with the resource. Maybe something like "ResourceInformation"...?

File: src/main/resources/graphql/query.graphqls (hunk 1 above)
Comment (myluki2000) [reply src/main/resources/graphql/query.graphqls:26]: In tandem with the name change I suggested for the Resource type, I'd suggest calling this query something like resourceInformationById

**Review by myluki2000 — CHANGES_REQUESTED** [review]

File: src/main/resources/graphql/resource.graphqls (hunk 4 above)
Comment (PaulBredl) [reply src/main/resources/graphql/resource.graphqls:15]: As already discussed with Tim, this type will be removed in favor of just an array of courses

File: src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (BlueWaves2) [reply src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java:23]: No particular reason. Just force of habit, lol

File: src/main/resources/graphql/resource.graphqls (hunk 5 above)
Comment (BlueWaves2) [reply src/main/resources/graphql/resource.graphqls:4]: I agree with this naming being suboptimal/misleading, but could not think of a better name to describe the Entity

File: src/main/resources/graphql/resource.graphqls [hunk 6]
```diff
@@ -5,20 +5,13 @@ type Resource {
...
     """
-    course_id: UUID!
+    A list of course IDs a resource is presently unavailable in
     """
-    If resource is available in course
-    """
-    available: Boolean!
-}
\ No newline at end of file
+    unAvailableCourses: [UUID!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/resource.graphqls:16]: What would the use case for this?

File: src/test/java/de/unistuttgart/iste/gits/courseservice/service/ResourceServiceTest.java [hunk 7]
```diff
@@ -128,7 +122,7 @@ private ResourceEntity.ResourceEntityBuilder dummyResourceEntityBuilder(UUID cou
         return ResourceEntity.builder().courseId(courseId).resourceId(resourceId);
     }
 
-    private ResourceDto.Builder dummyResourceDtoBuilder(UUID resourceId, List<CourseIdAvailabilityMapDto> courses){
-        return ResourceDto.builder().setResource_id(resourceId).setCourses(courses);
+    private ResourceDto.Builder dummyResourceDtoBuilder(UUID resourceId, List<UUID> available, List<UUID> unavailable){
+        return ResourceDto.builder().setId(resourceId).setAvailableCourses(available).setUnAvailableCourses(unavailable);
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/courseservice/service/ResourceServiceTest.java:126]: if you use a method to pass all arguments and then put them in the builder, that defeats the purpose of a builder and you can just call the constructor.
(that's not a todo, just a remark for the future)

File: src/test/java/de/unistuttgart/iste/gits/courseservice/integration/QueryResourcesTest.java [hunk 8]
```diff
@@ -35,17 +34,15 @@ void testResourceNotExisting(GraphQlTester tester){
...
-                    courseIdsByResource(ids: ["%s"]) {
-                        resource_id
-                        courses {
-                            course_id
-                            available
-                        }
+                    resourceById(ids: ["%s"]) {
+                        id
+                        availableCourses
+                        unAvailableCourses
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/courseservice/integration/QueryResourcesTest.java:40]: @myluki2000 do you think we need a list of "unavailable" course ids (i.e., the resource belongs to a course but the course is not started or public yet) -> I thought no, but Tim thought so, so maybe you have a better understanding than me?

File: src/main/resources/graphql/resource.graphqls [hunk 9]
```diff
@@ -5,20 +5,13 @@ type Resource {
...
-    A list of courses a resource is present in, including its availability
+    A list of course IDs a resource is presently available in
     """
-    courses: [courseIdAvailabilityMap!]!
-}
-
-type courseIdAvailabilityMap {
-    """
-    Course ID
+    availableCourses: [UUID!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/resource.graphqls:12]: would be nice to return real Course objects but just the UUID is also fine.

File: src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/courseservice/persistence/dao/ResourcePk.java:23]: I think you should either remove the data annotation or the explicit methods then IMO

**Review by PaulBredl — COMMENTED** [review]
I will wait for Lucas response before approving, all comments are optional TODOs

File: src/main/resources/graphql/resource.graphqls (hunk 5 above)
Comment (PaulBredl) [reply src/main/resources/graphql/resource.graphqls:4]: image

File: src/test/java/de/unistuttgart/iste/gits/courseservice/integration/QueryResourcesTest.java (hunk 8 above)
Comment (myluki2000) [reply src/test/java/de/unistuttgart/iste/gits/courseservice/integration/QueryResourcesTest.java:40]: I'm not sure if we need it, so I'd leave it in. "lieber haben und nicht brauchen als brauchen und nicht haben" as we say in German. Although I have 1 little nag: I think it should be named "unavailable" and not "unAvailable", as unavailable is one word.

Comment (PaulBredl) [reply src/test/java/de/unistuttgart/iste/gits/courseservice/integration/QueryResourcesTest.java:40]: I mean the counter argument would be that you don't want to maintain things that you don't really need, but I'm fine with leaving it in.

File: src/test/java/de/unistuttgart/iste/gits/courseservice/service/ResourceServiceTest.java (hunk 7 above)
Comment (BlueWaves2) [reply src/test/java/de/unistuttgart/iste/gits/courseservice/service/ResourceServiceTest.java:126]: I agree are the leftovers of a previous refactoring of the representation of the PK for the resource. I'll remove it

**Review by PaulBredl — APPROVED** [review]
lgtm

**Review by myluki2000 — APPROVED** [review]
